File: XcodeTemplates/Templates/Project Templates/iOS/Application Extension/Print Service Extension.xctemplate/PrintService.m
//___FILEHEADER___

#import "___FILEBASENAME___.h"

@implementation ___FILEBASENAME___

- (NSArray <UIPrinterDestination *> *)printerDestinationsForPrintInfo:(UIPrintInfo *)printInfo
{
    NSMutableArray <UIPrinterDestination *> *destinations = [NSMutableArray new];
    
    // Use an app group to obtain shared preferences. This App Group must also be specified in the app's entitlements.
    
    NSUserDefaults *groupDefaults = [[NSUserDefaults alloc] initWithSuiteName:@"group.com.example.PrintServiceApp"];
    NSDictionary *cloudPreferences = [groupDefaults dictionaryForKey:@"cloud_prefs"];
    
    if (cloudPreferences) {
        NSString *urlString = cloudPreferences[@"cloud_url"];
        
        NSURL *url = [NSURL URLWithString:urlString];
        if (url != nil) {
            UIPrinterDestination *destination = [[UIPrinterDestination alloc] initWithURL:url];
            
            destination.displayName = cloudPreferences[@"cloud_dpy"];
            
            NSDictionary <NSString *, NSString *> *txtDict = cloudPreferences[@"cloud_txt"];
            if (txtDict) {
                NSMutableDictionary<NSString*, NSData*> *txtDataDict = [NSMutableDictionary new];
                
                [txtDict enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
                    txtDataDict[key] = [obj dataUsingEncoding:NSASCIIStringEncoding];
                }];
                
                destination.txtRecord = [NSNetService dataFromTXTRecordDictionary:txtDataDict];
            }
            
            [destinations addObject:destination];
        }
    }
    
    return destinations;
}

@end

File: XcodeTemplates/Templates/Project Templates/iOS/Application Extension/Print Service Extension.xctemplate/PrintService.swift
//___FILEHEADER___

import UIKit

class ___FILEBASENAME___: UIPrintServiceExtension {
    
    override func printerDestinations(for printInfo: UIPrintInfo) -> [UIPrinterDestination] {
        var destinations: [UIPrinterDestination] = []
        
        // Use an app group to obtain shared preferences. Replace the suite name with the one in your app's entitlements.
        
        let groupDefaults = UserDefaults(suiteName: "group.com.example.PrinterServiceSampleApp")
        
        if let cloudPreferences = groupDefaults?.dictionary(forKey: "cloud_prefs"),
           let urlString = cloudPreferences["cloud_url"] as? String,
           let url = URL(string: urlString)
        {
            let destination = UIPrinterDestination(url: url)
            
            destination.displayName = cloudPreferences["cloud_dpy"] as? String
            
            if let txtDict = cloudPreferences["cloud_txt"] as? Dictionary<String, String> {
                let txtDataDict: [String: Data] = txtDict.compactMapValues { $0.data(using: .ascii) }
                
                destination.txtRecord = NetService.data(fromTXTRecord: txtDataDict)
            }
            
            destinations.append(destination)
        }
        
        return destinations
    }
    
}
